jujusvg.go

```diff
@@ -13,8 +13,33 @@ import (
 // NewFromBundle returns a new Canvas that can be used
 // to generate a graphical representation of the given bundle
 // data. The iconURL function is used to generate a URL
-// that refers to an SVG for the supplied charm URL.
-func NewFromBundle(b *charm.BundleData, iconURL func(*charm.Reference) string) (*Canvas, error) {
+// that refers to an SVG for the supplied charm URL.  If
+// requested, icon SVGs that are available will be embedded
+// within the bundle SVG so that the bundle will be self-
+// contained as much as possible.  If that is to be the case,
+// utilize the provided IconFetcher (or default to HTTPFetcher
+// if nil) to retrieve the icons.
+func NewFromBundle(b *charm.BundleData, iconURL func(*charm.Reference) string, fetcher IconFetcher) (*Canvas, error) {
+	if fetcher == nil {
+		fetcher = &LinkFetcher{
+			IconURL: iconURL,
+		}
+	}
+	iconMap, err := fetcher.FetchIcons(b)
+	if err != nil {
+		return nil, err
+	}
+	return newFromBundleWithMap(b, iconURL, iconMap)
+}
+
+// newFromBundleWithMap returns a new Canvas that can be used
+// to generate a graphical representation of the given bundle
+// data. The iconURL function is used to generate a URL
+// that refers to an SVG for the supplied charm URL. If a map
+// of charms to icon SVGs is provided, then those SVGs will be
+// embedded in the bundle diagram and used instead of an image
+// tag.
+func newFromBundleWithMap(b *charm.BundleData, iconURL func(*charm.Reference) string, iconMap map[string][]byte) (*Canvas, error) {
```

Comment:
This function can probably be merged into NewWithBundle now without a decrease in readability.